From 8c8d6faa9f0fc35581e95a9e771490feb94dd9ac Mon Sep 17 00:00:00 2001
From: Sebastian Wick <sebastian.wick@redhat.com>
Date: Tue, 16 Dec 2025 17:15:32 +0100
Subject: [PATCH] run: Enable FIPS crypto policy if it is enabled on the host

This is a close copy of what podman/containers does to support FIPS. Any
other crypto policy is ignored for now.
---
 common/flatpak-run.c | 94 ++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 94 insertions(+)

diff --git a/common/flatpak-run.c b/common/flatpak-run.c
index dfcef03..121bef3 100644
--- a/common/flatpak-run.c
+++ b/common/flatpak-run.c
@@ -2214,6 +2214,98 @@ flatpak_run_setup_usr_links (FlatpakBwrap *bwrap,
     }
 }
 
+static void
+flatpak_run_setup_fips (FlatpakBwrap *bwrap,
+                        int           runtime_fd)
+{
+  glnx_autofd int crypto_policies_fd = -1;
+  glnx_autofd int fips_backend_fd = -1;
+  glnx_autofd int fips_config_fd = -1;
+  g_autofree char *fips_enabled = NULL;
+  g_autoptr(GError) error = NULL;
+
+  if (runtime_fd < 0)
+    return;
+
+  if (!g_file_get_contents ("/proc/sys/crypto/fips_enabled",
+                            &fips_enabled,
+                            NULL, &error))
+    {
+      if (!g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_NOENT))
+        {
+          g_warning ("Failed to read /proc/sys/crypto/fips_enabled to determine FIPS state: %s",
+                     error->message);
+        }
+
+      return;
+    }
+
+  g_strstrip (fips_enabled);
+
+  if (g_strcmp0 (fips_enabled, "1") != 0)
+    {
+      g_info ("FIPS is disabled");
+      return;
+    }
+
+  crypto_policies_fd = glnx_chaseat (runtime_fd, "etc/crypto-policies",
+                                     GLNX_CHASE_RESOLVE_BENEATH,
+                                     NULL);
+
+  if (crypto_policies_fd < 0)
+    {
+      g_info ("FIPS is enabled, but runtime does not support it");
+      return;
+    }
+
+  fips_backend_fd = glnx_chaseat (runtime_fd,
+                                   "share/crypto-policies/back-ends/FIPS",
+                                   GLNX_CHASE_RESOLVE_BENEATH,
+                                   NULL);
+
+  if (fips_backend_fd < 0)
+    {
+      g_info ("FIPS is enabled, but runtime does not support it");
+      return;
+    }
+
+  fips_config_fd = glnx_chaseat (runtime_fd,
+                                  "share/crypto-policies/default-fips-config",
+                                  GLNX_CHASE_RESOLVE_BENEATH |
+                                  GLNX_CHASE_MUST_BE_REGULAR,
+                                  NULL);
+
+  if (fips_config_fd >= 0)
+    {
+      flatpak_bwrap_add_args_data_fd (bwrap,
+                                      "--ro-bind-fd",
+                                      g_steal_fd (&fips_config_fd),
+                                      "/etc/crypto-policies/config");
+    }
+  else
+    {
+      if (!flatpak_bwrap_add_args_data (bwrap,
+                                        "default-fips-config",
+                                        "FIPS\n",
+                                        -1,
+                                        "/etc/crypto-policies/config",
+                                        &error))
+        {
+          g_warning ("Failed to enable FIPS configuration: "
+                     "creating default-fips-config tmpfile failed: %s",
+                     error->message);
+          return;
+        }
+    }
+
+  flatpak_bwrap_add_args_data_fd (bwrap,
+                                  "--ro-bind-fd",
+                                  g_steal_fd (&fips_backend_fd),
+                                  "/etc/crypto-policies/back-ends");
+
+  g_info ("Enabled FIPS configuration");
+}
+
 /* Directories in /sys to share with the sandbox if accessible. */
 static const char *const sysfs_dirs[] =
 {
@@ -2426,6 +2518,8 @@ flatpak_run_setup_base_argv (FlatpakBwrap   *bwrap,
         }
     }
 
+  flatpak_run_setup_fips (bwrap, runtime_fd);
+
   if (app_id_dir != NULL)
     {
       g_autoptr(GFile) app_cache_dir = g_file_get_child (app_id_dir, "cache");
-- 
2.54.0

